File: shared.js
/*******************************************************
 * Author: Matthew Taylor
 * Created: 2024
 *******************************************************/

const WINNING_MASKS = [
    0b000_000_111,  // Row 0
    0b000_111_000,  // Row 1
    0b111_000_000,  // Row 2
    0b001_001_001,  // Col 0
    0b010_010_010,  // Col 1
    0b100_100_100,  // Col 2
    0b100_010_001,  // Diag 1
    0b001_010_100,  // Diag 2
];

class GameConfig {
    constructor(p0_s, p0_m, p0_l, p1_s, p1_m, p1_l, p0_unused, p1_unused) {
        this.p0_s = p0_s; this.p0_m = p0_m; this.p0_l = p0_l;
        this.p1_s = p1_s; this.p1_m = p1_m; this.p1_l = p1_l;
        this.p0_unused = p0_unused; // [small, medium, large]
        this.p1_unused = p1_unused; // [small, medium, large]
    }
}

class Move {
    constructor(size, start_sq, end_sq) {
        this.size = size;
        this.start_sq = start_sq; // null if from unused
        this.end_sq = end_sq;
    }
}

function getWinner(game) {
    const l_any = game.p0_l | game.p1_l;
    const m_any = game.p0_m | game.p1_m;

    const vis_0_l = game.p0_l;
    const vis_0_m = game.p0_m & ~l_any;
    const vis_0_s = game.p0_s & ~l_any & ~m_any;
    const p0_control = vis_0_l | vis_0_m | vis_0_s;

    for (let mask of WINNING_MASKS) {
        if ((p0_control & mask) === mask) return 0;
    }

    const vis_1_l = game.p1_l;
    const vis_1_m = game.p1_m & ~l_any;
    const vis_1_s = game.p1_s & ~l_any & ~m_any;
    const p1_control = vis_1_l | vis_1_m | vis_1_s;

    for (let mask of WINNING_MASKS) {
        if ((p1_control & mask) === mask) return 1;
    }

    return null;
}

function getTopPiece(game, sq) {
    const mask = 1 << sq;
    if (game.p0_l & mask) return { color: 0, size: 2 };
    if (game.p1_l & mask) return { color: 1, size: 2 };
    if (game.p0_m & mask) return { color: 0, size: 1 };
    if (game.p1_m & mask) return { color: 1, size: 1 };
    if (game.p0_s & mask) return { color: 0, size: 0 };
    if (game.p1_s & mask) return { color: 1, size: 0 };
    return null;
}

function getBits(mask) {
    const bits = [];
    for (let i = 0; i < 9; i++) {
        if (mask & (1 << i)) bits.push(i);
    }
    return bits;
}

function getPossibleMoves(game, player) {
    const moves = [];
    if (getWinner(game) !== null) return moves;

    const unused = player === 0 ? game.p0_unused : game.p1_unused;
    const l_any = game.p0_l | game.p1_l;
    const m_any = game.p0_m | game.p1_m;
    const s_any = game.p0_s | game.p1_s;

    // JavaScript bitwise operations are 32-bit signed, restrict inversion to our 9-bit board layout
    const can_place_l = (~l_any) & 0x1FF;
    const can_place_m = (~(l_any | m_any)) & 0x1FF;
    const can_place_s = (~(l_any | m_any | s_any)) & 0x1FF;

    if (unused[2] > 0) {
        for (let sq of getBits(can_place_l)) moves.push(new Move(2, null, sq));
    }
    if (unused[1] > 0) {
        for (let sq of getBits(can_place_m)) moves.push(new Move(1, null, sq));
    }
    if (unused[0] > 0) {
        for (let sq of getBits(can_place_s)) moves.push(new Move(0, null, sq));
    }

    const vis_l = player === 0 ? game.p0_l : game.p1_l;
    const vis_m = (player === 0 ? game.p0_m : game.p1_m) & ~l_any;
    const vis_s = (player === 0 ? game.p0_s : game.p1_s) & ~(l_any | m_any);

    for (let sq of getBits(vis_l)) {
        for (let target of getBits(can_place_l)) {
            if (target !== sq) moves.push(new Move(2, sq, target));
        }
    }
    for (let sq of getBits(vis_m)) {
        for (let target of getBits(can_place_m)) {
            if (target !== sq) moves.push(new Move(1, sq, target));
        }
    }
    for (let sq of getBits(vis_s)) {
        for (let target of getBits(can_place_s)) {
            if (target !== sq) moves.push(new Move(0, sq, target));
        }
    }

    return moves;
}

function nextPlayer(player) {
    return 1 - player;
}

function playMove(game, player, move) {
    let p0_s = game.p0_s, p0_m = game.p0_m, p0_l = game.p0_l;
    let p1_s = game.p1_s, p1_m = game.p1_m, p1_l = game.p1_l;
    let p0_u = [...game.p0_unused], p1_u = [...game.p1_unused];

    const { size, start_sq, end_sq } = move;

    if (start_sq !== null) {
        const mask = ~(1 << start_sq);
        if (player === 0) {
            if (size === 0) p0_s &= mask;
            else if (size === 1) p0_m &= mask;
            else p0_l &= mask;
        } else {
            if (size === 0) p1_s &= mask;
            else if (size === 1) p1_m &= mask;
            else p1_l &= mask;
        }
    } else {
        if (player === 0) p0_u[size] -= 1;
        else p1_u[size] -= 1;
    }

    const target_mask = 1 << end_sq;
    if (player === 0) {
        if (size === 0) p0_s |= target_mask;
        else if (size === 1) p0_m |= target_mask;
        else p0_l |= target_mask;
    } else {
        if (size === 0) p1_s |= target_mask;
        else if (size === 1) p1_m |= target_mask;
        else p1_l |= target_mask;
    }

    return new GameConfig(p0_s, p0_m, p0_l, p1_s, p1_m, p1_l, p0_u, p1_u);
}

function getScore(game, player) {
    const winner = getWinner(game);
    if (winner === player) return 1000;
    if (winner === nextPlayer(player)) return -1000;
    return 0;
}

function alphabetaMinNode(board, color, alpha, beta, limit) {
    const nextCol = nextPlayer(color);
    const moves = getPossibleMoves(board, nextCol);
    let bestMove = null;
    let bestUtility = Infinity;

    if (limit <= 0 || moves.length === 0) {
        return [null, getScore(board, color)];
    }

    for (let move of moves) {
        const newBoard = playMove(board, nextCol, move);
        const [_, utility] = alphabetaMaxNode(newBoard, color, alpha, beta, limit - 1);

        if (utility < bestUtility) {
            bestMove = move;
            bestUtility = utility;
        }

        beta = Math.min(beta, utility);
        if (beta <= alpha) break;
    }

    return [bestMove, bestUtility];
}

function alphabetaMaxNode(board, color, alpha, beta, limit) {
    const moves = getPossibleMoves(board, color);
    let bestMove = null;
    let bestUtility = -Infinity;

    if (limit <= 0 || moves.length === 0) {
        return [null, getScore(board, color)];
    }

    for (let move of moves) {
        const newBoard = playMove(board, color, move);
        const [_, utility] = alphabetaMinNode(newBoard, color, alpha, beta, limit - 1);

        if (utility > bestUtility) {
            bestMove = move;
            bestUtility = utility;
        }

        alpha = Math.max(alpha, utility);
        if (beta <= alpha) break;
    }

    return [bestMove, bestUtility];
}

function selectMoveAlphabeta(gameConfig, color, limit) {
    return alphabetaMaxNode(gameConfig, color, -Infinity, Infinity, limit);
}

File: gobblet_py/latest/shared.py
"""
 * Author: Matthew Taylor
 * Created: 2024
"""

from typing import NamedTuple

WINNING_MASKS = (
    0b000_000_111,  # Row 0
    0b000_111_000,  # Row 1
    0b111_000_000,  # Row 2
    0b001_001_001,  # Col 0
    0b010_010_010,  # Col 1
    0b100_100_100,  # Col 2
    0b100_010_001,  # Diag 1
    0b001_010_100,  # Diag 2
)

# Player 0:
# bits  0-1   small
# bits  2-3   medium
# bits  4-5   large

# Player 1:
# bits  6-7   small
# bits  8-9   medium
# bits 10-11  large

INITIAL_UNUSED = (
    (2 << 0)  |  # p0 small
    (2 << 2)  |  # p0 medium
    (2 << 4)  |  # p0 large
    (2 << 6)  |  # p1 small
    (2 << 8)  |  # p1 medium
    (2 << 10)    # p1 large
)

class GameConfig(NamedTuple):
    # Bitboards for player 0 (bits 0-8 represent squares left-to-right, top-to-bottom)
    p0_s: int
    p0_m: int
    p0_l: int

    # Bitboards for player 1
    p1_s: int
    p1_m: int
    p1_l: int

    # Counts of unused pieces: 
    # (small, medium, large) (small, medium, large) for both players, packed into int
    unused: int


class Move(NamedTuple):
    size: int  # 0 = small, 1 = medium, 2 = large
    start_sq: int | None  # 0-8 for board positions, None if placing from unused
    end_sq: int  # 0-8 for destination


def game_over(game: GameConfig) -> int | None:
    l_any = game.p0_l | game.p1_l
    m_any = game.p0_m | game.p1_m

    p0_control = (
        game.p0_l |
        (game.p0_m & ~l_any) |
        (game.p0_s & ~l_any & ~m_any)
    )
    p1_control = (
        game.p1_l |
        (game.p1_m & ~l_any) |
        (game.p1_s & ~l_any & ~m_any)
    )

    return any(
        (p0_control & mask) == mask or
        (p1_control & mask) == mask
        for mask in WINNING_MASKS
    )


def get_top_piece(game: GameConfig, sq: int) -> tuple[int | None, int | None]:
    """Returns (color, size) of the top piece on a given square, or (None, None)."""
    mask = 1 << sq
    if game.p0_l & mask:
        return 0, 2
    if game.p1_l & mask:
        return 1, 2
    if game.p0_m & mask:
        return 0, 1
    if game.p1_m & mask:
        return 1, 1
    if game.p0_s & mask:
        return 0, 0
    if game.p1_s & mask:
        return 1, 0
    return None, None


def get_bits(mask: int):
    for i in range(9):
        if mask & (1 << i):
            yield i
            
            

def get_unused(unused: int, player: int, size: int) -> int:
    shift = player * 6 + size * 2
    return (unused >> shift) & 0b11

def dec_unused(unused: int, player: int, size: int) -> int:
    return unused - (1 << (player * 6 + size * 2))


def get_possible_moves(game: GameConfig, player: int) -> list[Move]:
    moves: list[Move] = []

    # Combined bitmasks of all pieces by size across both players
    l_any = game.p0_l | game.p1_l
    m_any = game.p0_m | game.p1_m
    s_any = game.p0_s | game.p1_s

    # Squares available for placing pieces of various sizes
    can_place_l = ~l_any & 0x1FF
    can_place_m = ~(l_any | m_any) & 0x1FF
    can_place_s = ~(l_any | m_any | s_any) & 0x1FF

    # Placing unused pieces
    if get_unused(game.unused, player, 0):
        for sq in get_bits(can_place_s):
            moves.append(Move(0, None, sq))
    if get_unused(game.unused, player, 1) > 0:
        for sq in get_bits(can_place_m):
            moves.append(Move(1, None, sq))
    if get_unused(game.unused, player, 2) > 0:
        for sq in get_bits(can_place_l):
            moves.append(Move(2, None, sq))

    # Moving visible pieces on the board
    vis_l = game.p0_l if player == 0 else game.p1_l
    vis_m = (game.p0_m if player == 0 else game.p1_m) & ~l_any
    vis_s = (game.p0_s if player == 0 else game.p1_s) & ~(l_any | m_any)
    
    for sq in get_bits(vis_s):
        for target in get_bits(can_place_s):
            if target != sq:
                moves.append(Move(0, sq, target))
                
    for sq in get_bits(vis_m):
        for target in get_bits(can_place_m):
            if target != sq:
                moves.append(Move(1, sq, target))

    for sq in get_bits(vis_l):
        for target in get_bits(can_place_l):
            if target != sq:
                moves.append(Move(2, sq, target))

    return moves


def next_player(player: int) -> int:
    return 1 - player


def play_move(game: GameConfig, player: int, move: Move) -> GameConfig:
    boards = [
        game.p0_s, game.p0_m, game.p0_l,
        game.p1_s, game.p1_m, game.p1_l,
    ]

    size, start_sq, end_sq = move
    idx = player * 3 + size
    unused = game.unused

    if start_sq is None:
        unused = dec_unused(game.unused, player, size)
    else:
        boards[idx] &= ~(1 << start_sq)

    boards[idx] |= 1 << end_sq

    return GameConfig(*boards, unused)
    
    
def get_winner(game: GameConfig) -> int | None:
    l_any = game.p0_l | game.p1_l
    m_any = game.p0_m | game.p1_m

    p0_control = (
        game.p0_l |
        (game.p0_m & ~l_any) |
        (game.p0_s & ~l_any & ~m_any)
    )
    p1_control = (
        game.p1_l |
        (game.p1_m & ~l_any) |
        (game.p1_s & ~l_any & ~m_any)
    )
    
    p0_wins = any((p0_control & mask) == mask for mask in WINNING_MASKS)
    p1_wins = any((p1_control & mask) == mask for mask in WINNING_MASKS)
    
    if p0_wins and p1_wins:
        return -1  # Tie
    if p0_wins:
        return 0
    if p1_wins:
        return 1
        
    return None


def get_control(game: GameConfig, player: int):
    l_any = game.p0_l | game.p1_l
    m_any = game.p0_m | game.p1_m

    if player == 0:
        return (
            game.p0_l |
            (game.p0_m & ~l_any) |
            (game.p0_s & ~l_any & ~m_any)
        )

    else:
        return (
            game.p1_l |
            (game.p1_m & ~l_any) |
            (game.p1_s & ~l_any & ~m_any)
        )


def get_score(game: GameConfig, player: int) -> int:
    winner = get_winner(game)

    if winner == player:
        return 100000

    score = 0
    control = get_control(game, player)

    for mask in WINNING_MASKS:
        pieces = (control & mask).bit_count()

        if pieces == 2:
            score += 100

    return score


def move_score(game: GameConfig, player: int, move: Move) -> int:
    # Score moves based on how many lines they create or block
    new_game = play_move(game, player, move)
    return get_score(new_game, player)



WIN_SCORE = 1000000

def negamax(game, player, alpha, beta, limit):
    winner = get_winner(game)

    if winner is not None:
        if winner == -1:
            return 0, None
        elif winner == player:
            # add a bonus for winning sooner
            return WIN_SCORE + limit, None
        else:
            return -WIN_SCORE - limit, None
    
    if limit <= 0:
        return get_score(game, player), None

    moves = get_possible_moves(game, player)

    if not moves:
        return 0, None

    moves.sort(
        key=lambda m: move_score(game, player, m),
        reverse=True,
    )

    best_move = None
    best_score = -float("inf")

    for move in moves:
        child = play_move(game, player, move)
        score, _ = negamax(child, 1-player, -beta, -alpha, limit - 1)
        score = -score

        if score > best_score:
            best_score = score
            best_move = move
 
        alpha = max(alpha, score)

        if alpha >= beta:
            break

    return best_score, best_move


def select_move_alphabeta(game_config: GameConfig, color, limit):
    return negamax(game_config, color, -float("inf"), float("inf"), limit)
